Source/Particles/Deposition/ChargeDeposition.H

```diff
@@ -201,6 +202,302 @@ void doChargeDepositionShapeN (const GetParticlePosition& GetPosition,
+
+    const amrex::Box sample_tbox(IntVect(AMREX_D_DECL(0, 0, 0)), a_tbox_max_size - 1);
+    amrex::Box sample_tbox_x = convert(sample_tbox, ix_type);
+
+    sample_tbox_x.grow(depos_order);
+
+    const auto npts = sample_tbox_x.numPts();
+
+    const int nblocks = a_bins.numBins();
+    const auto offsets_ptr = a_bins.offsetsPtr();
+    const int threads_per_block = 256;
+
+    std::size_t shared_mem_bytes = npts*sizeof(amrex::Real);
+    amrex::IntVect bin_size = WarpX::shared_tilesize;
+
+    const std::size_t max_shared_mem_bytes = amrex::Gpu::Device::sharedMemPerBlock();
+
+    WARPX_ALWAYS_ASSERT_WITH_MESSAGE(shared_mem_bytes <= max_shared_mem_bytes,
+                                      "Tile size too big for GPU shared memory charge deposition");
+#endif
+
+#if defined(AMREX_USE_CUDA) || defined(AMREX_USE_HIP)
+    amrex::ignore_unused(np_to_deposit);
+    // Loop with one block per tile (the shared memory is allocated on a per-block basis)
+    // The threads within each block loop over the particles of its tile
+    // (Each threads processes a different set of particles.)
+    amrex::launch(
+                  nblocks, threads_per_block, shared_mem_bytes, amrex::Gpu::gpuStream(),
+                  [=] AMREX_GPU_DEVICE () noexcept
+#else // defined(AMREX_USE_CUDA) || defined(AMREX_USE_HIP)
+    amrex::ParallelFor(np_to_deposit, [=] AMREX_GPU_DEVICE (long ip_orig) noexcept
+#endif
+      {
+#if defined(AMREX_USE_CUDA) || defined(AMREX_USE_HIP)
+        const int bin_id = blockIdx.x;
+        const unsigned int bin_start = offsets_ptr[bin_id];
+        const unsigned int bin_stop = offsets_ptr[bin_id+1];
+
+        if (bin_start == bin_stop) { return; }
+
+        amrex::Box buffer_box;
+        {
+          ParticleReal xp, yp, zp;
+          GetPosition(permutation[bin_start], xp, yp, zp);
+#if defined(WARPX_DIM_3D)
+          IntVect iv = IntVect(int(amrex::Math::floor((xp-plo[0])*dxiarr[0])),
+                               int(amrex::Math::floor((yp-plo[1])*dxiarr[1])),
+                               int(amrex::Math::floor((zp-plo[2])*dxiarr[2])));
+#elif defined(WARPX_DIM_XZ) || defined(WARPX_DIM_RZ)
+          IntVect iv = IntVect(
+                               int(amrex::Math::floor((xp-plo[0])*dxiarr[0])),
+                               int(amrex::Math::floor((zp-plo[1])*dxiarr[1])));
+#elif defined(WARPX_DIM_1D_Z)
+          IntVect iv = IntVect(int(amrex::Math::floor((zp-plo[0])*dxiarr[0])));
+#endif
+          iv += domain.smallEnd();
+          getTileIndex(iv, box, true, bin_size, buffer_box);
+        }
+
+        Box tbx = convert( buffer_box, ix_type);
+        tbx.grow(depos_order);
+
+        Gpu::SharedMemory<amrex::Real> gsm;
+        amrex::Real* const shared = gsm.dataPtr();
+
+        amrex::Array4<amrex::Real> buf(shared, amrex::begin(tbx), amrex::end(tbx), 1);
+
+        // Zero-initialize the temporary array in shared memory
+        volatile amrex::Real* vs = shared;
```

Comment:
Why do we need `volatile` here?
(I remember an ancient CUDA quirk with multiple shared memory locations that needed this.)

Reply:
I think you're right that we technically don't need it. From what I understand, the `volatile` stops compiler optimizations from storing shared values in registers, which can generally cause problems when threads are accessing the same value in shared memory, as we do. I think because of the places we `__syncthreads`, and because most of our accesses to shared are atomics, we should be ok, but it's the kind of thing we'd need to make absolute sure of, and even a slight change later on, like moving the `__syncthreads`, could lead to some inexplicable bugs.  I ran some tests, and didn't see any performance benefit from removing it, so I'd recommend keeping it.